### PR TITLE
[Ready to Review] Incorrect permissions and links for node links 

### PR DESCRIPTION
**Pull request by pattisdr**
# Purpose
Issue OSF-4630

The Node Link serialization isn't correct:
"self" link should point to the Node Link Detail URL, not the target node's URL.
The target node's url should go in /data/relationships/node/links/related/href

The DELETE behavior for NodeLinkDetail is also incorrect:
The user only needs permission to remove the node link; not edit permissions for the target node
Note: Node link endpoints are currently dev-only, but this is a blocker for the public stable release.

# Changes
screen shot 2015-10-08 at 10 19 25 am

This PR 1) makes self link pointing to Node Link Detail URL 2) moves target node's url under relationships 3)  Changes permissions for deleting node-link 4) For now, makes target_node_id write-only.  

Separate PR https://openscience.atlassian.net/browse/OSF-4804 will change how node links are created.

### CODE REVIEW

**Comment by sloria** [comment]
Looks good. CR1 finished.

File: tests/api_tests/nodes/test_views.py [hunk 1]
```diff
@@ -2666,7 +2670,8 @@ def test_creates_private_node_pointer_logged_out(self):
     def test_creates_private_node_pointer_logged_in_contributor(self):
         res = self.app.post_json_api(self.private_url, self.private_payload, auth=self.user.auth)
         assert_equal(res.status_code, 201)
-        assert_equal(res.json['data']['attributes']['target_node_id'], self.pointer_project._id)
+        url = res.json['data']['relationships']['target_node']['links']['related']['href']
+        assert_in('/{}nodes/{}/'.format(API_BASE, self.pointer_project._id), url)
```
Comment (abought) [inline tests/api_tests/nodes/test_views.py]: This pattern is copied many times across the tests in this file. To increase maintainability if the URL format ever changes ever change, consider defining all the duplicate code in one place (DRY) and using it as follows.

```python
# Before any tests
node_url_for = lambda n_id: '/{}nodes/{}/'.format(API_BASE, n_id)
...
def test_method(self, ...):
    ...
    expected_url = node_url_for(self.pointer_project._id)
    assert_in(expected_url, url)
```

```

File: tests/api_tests/nodes/test_views.py [hunk 2]
```diff
@@ -2447,8 +2447,9 @@ def test_return_public_node_pointers_logged_out(self):
         res_json = res.json['data']
         assert_equal(len(res_json), 1)
         assert_equal(res.status_code, 200)
-        assert_in(res_json[0]['attributes']['target_node_id'], self.public_pointer_project._id)
         assert_equal(res.content_type, 'application/vnd.api+json')
+        url = res.json['data'][0]['relationships']['target_node']['links']['related']['href']
+        assert_in('/{}nodes/{}/'.format(API_BASE, self.public_pointer_project._id), url)
```
Comment (abought) [inline tests/api_tests/nodes/test_views.py]: Can you please clarify why these tests use `assert_in` instead of `assert_equal`?

Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: I'm asserting that the first is in the second.  For some reason, the url being returned here has localhost:80 instead of 8000 so I just comare the /v2/nodes/pgvjg.  Outside of the tests, the url is returned correctly localhost:8000/...

screen shot 2015-10-14 at 6 46 54 pm

Comment (abought) [inline tests/api_tests/nodes/test_views.py]: The port difference isn't too surprising- after all, webtest doesn't run by making requests to your `apiserver`. (if you don't believe me, try running your tests without `apiserver` running- [just like Travis](https://github.com/abought/osf.io/blob/develop/.travis.yml#L88-88)!)

Webtest internally has its own system of requests for testing, which appears to [default to port 80](http://stackoverflow.com/questions/2774249/python-webtest-port-configuration). You could change it, but really not worth the headache.

Comment (abought) [inline tests/api_tests/nodes/test_views.py]: As for `assert_in`, all of these tests using `assert_in` seem fragile, because they would would pass even if the API gave back a different URL like `http://localhost:80/v2/nodes/pgvjq/contributors/`.

What you really want is to ensure the end part of the url (after `localhost:port`) is the same (and that there aren't any other bits). The below suggestion is a hack at 11:30pm; there may be a more elegant option.

```
url_matches = actual_url.endswith(expected_url)
assert_true(url_matches)
```

Since there are a lot of tests that follow this pattern, a lot of tests will need to be updated. (This is where helper methods that DRY common code come in handy!)

File: tests/api_tests/nodes/test_views.py [hunk 3]
```diff
@@ -3079,7 +3095,14 @@ def setUp(self):
         self.public_pointer = self.public_project.add_pointer(self.public_pointer_project,
                                                               auth=Auth(self.user),
                                                               save=True)
-        self.public_url = '/{}nodes/{}/node_links/{}'.format(API_BASE, self.public_project._id, self.public_pointer._id)
+        self.public_url = '/{}nodes/{}/node_links/{}/'.format(API_BASE, self.public_project._id, self.public_pointer._id)
+
+    def test_delete_node_link_no_permissions_for_target_node(self):
+        pointer_project = ProjectFactory(creator=self.user_two, is_public=False)
+        pointer = self.public_project.add_pointer(pointer_project, auth=Auth(self.user), save=True)
```
Comment (abought) [inline tests/api_tests/nodes/test_views.py]: Instead of `Auth(self.user)`, you can use `self.user.auth` (since `self.user = AuthUserFactory()`)

File: tests/api_tests/nodes/test_views.py [hunk 4]
```diff
@@ -3079,7 +3095,14 @@ def setUp(self):
...
+        url = '/{}nodes/{}/node_links/{}/'.format(API_BASE, self.public_project._id, pointer._id)
+        res = self.app.delete_json_api(url, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 204)
```
Comment (abought) [inline tests/api_tests/nodes/test_views.py]: In addition to checking status code, you can optionally reload the project and check that the actual expected result (removal of link) has occurred.

File: api/nodes/serializers.py [hunk 5]
```diff
@@ -272,23 +273,31 @@ class NodeLinksSerializer(JSONAPISerializer):
 
     id = IDField(source='_id', read_only=True)
     type = TypeField()
-    target_node_id = ser.CharField(source='node._id', help_text='The ID of the node that this Node Link points to')
+    target_node_id = ser.CharField(write_only=True, source='node._id', help_text='The ID of the node that this Node Link points to')
```
Comment (abought) [inline api/nodes/serializers.py]: This change is mentioned, but no rationale is provided. What is this change intended to address?

**Comment by abought** [comment]
CR2 pass complete. The basic logic looks good, but there are a few test cases that seem ambiguous wrt pass/failure reasons. A few comments noted, but all pretty straightforward.

Adding some helper functions to clean up redundant operations (DRY) will make it much easier to update those tests, so I recommend a bit of cleanup there as well.

File: api/nodes/serializers.py (hunk 5 above)
Comment (pattisdr) [inline api/nodes/serializers.py]: Target node info doesn't need to be in two places.  Here, I need target_node_id to create the relationship, but I am returning the target_node information under relationships. In the other PR mentioned in this ticket, target_node_id won't be under attributes in the request either,  but will go under relationships.  It will be under relationships in the request, and under relationships in the response, as soon as both PR's are merged.

Comment (abought) [inline api/nodes/serializers.py]: That makes sense. Thanks.

File: tests/api_tests/nodes/test_views.py (hunk 3 above)
Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: @abought `Auth(self.user)` is different from `self.user.auth`. `Auth(self.user)` is an object that encapsulates identity information. `self.user.auth` is just a tuple with Basic Auth credentials, typically passed to webtest_plus.

Comment (abought) [inline tests/api_tests/nodes/test_views.py]: Whoops, thanks Steve. Today's lesson: I shouldn't suggest auth changes so soon after a big lunch.

**Comment by abought** [comment]
CR2 complete.

For future reference, it would be possible to DRY these tests quite a bit farther, but that's not something I'll push for in this PR; don't want to hold up a blocker.
